cloud/terraform/terraform_controller.py

```diff
@@ -0,0 +1,122 @@
+        if os.path.exists("main.tf"):
+            os.remove("main.tf")
+        header = """
+terraform {
+    required_providers {
+        aws = {
+        source  = "hashicorp/aws"
+        version = "~> 3.27"
+        }
+    }
+
+    required_version = ">= 0.14.9"
+}
+"""
+        self._add_to_config_file(header)
+
+    def _add_to_config_file(self, addition):
+        with open("main.tf", "a") as file:
+            file.write(addition)
+        
+    def _string_in_config(self, test_string):
+        with open("main.tf", "r") as file:
+            return (test_string in file.read())
+
+    # Credentials can be passed to terraform by adding the to the provider block
+    def aws_credentials(self, access_key, secret_key, token):
+        """
+        Give terraform controller aws credentials used on the configuration
+        """
+        self.aws_access_key = access_key
+        self.aws_secret_key = secret_key
+        self.token = token
+
+    def add_region(self, region):
+        """
+        Add region to configuration
+        """
+        if self.cloud == "aws":
+            if not self.aws_access_key:
+                print("Before adding any aws region, aws_credentials should be called")
+                exit()
+
+            region_block = """
+provider "aws" {{
+  alias   = "{region}"
+  region  = "{region}"
+  access_key = "{access_key}"
+  secret_key = "{secret_key}"
+  token = "{token}"
+}}
+""".format(region=region, access_key=self.aws_access_key, secret_key=self.aws_secret_key, token=self.token)
+            self._add_to_config_file(region_block)
+        
+    def add_instance(self, region, image_id, instance_type=""):
+        """
+        Add instance to configuration and returns the terraform resource id
+        """
+        if not self._string_in_config(region):
+            print("Cannot add an instance if region provider is not set up")
+            exit()
+        
+        if self.cloud == "aws":
+            if not instance_type:
+                instance_type = "t2.micro"
+            instance_block ="""
+resource "aws_instance" "instance-{region}" {{
+  provider      = aws.{region}
+  ami           = "{image_id}"
+  instance_type = "{instance_type}"
+  tags = {{
+    Name = "instance-{region}"
+  }}
+}}           
+""".format(region=region, image_id=image_id, instance_type=instance_type)
+            self._add_to_config_file(instance_block)
+            return f"aws_instance.instance-{region}"
+
+    def create_infra(self):
+        """
+        Create the infrastucture specified in the terraform configuration
+        """
+        cmd_output = os.system("terraform init")
+        if cmd_output:
+            print("terraform init command failed, check configuration")
+            exit()
+
+        cmd_output = os.system("terraform apply")
+        if cmd_output:
+            print("terraform apply command failed, check configuration")
+            exit()
+
+    def destroy_resource(self, resource_id):
```

Comment:
Question: Is the `resource_id` the same as the resource name?

Reply:
Nope, resource id is `<RESOURCE TYPE>.<NAME>`. the function add_instance returns the resource_id